lychee-lib/src/test_utils.rs

```diff
@@ -1,45 +1,45 @@
-#![allow(unused)]
-use http::StatusCode;
-use reqwest::Url;
-use wiremock::matchers::path;
-use wiremock::{Mock, MockServer, ResponseTemplate};
+use std::convert::TryFrom;
 
-use crate::Uri;
+use reqwest::Url;
 
-pub async fn get_mock_server<S>(response_code: S) -> MockServer
-where
-    S: Into<StatusCode>,
-{
-    get_mock_server_with_content(response_code, None).await
+use crate::{ClientBuilder, ErrorKind, Request, Uri};
+
+#[macro_export]
+macro_rules! mock_server {
+    ($status:expr $(, $func:tt ($($arg:expr),*))*) => {{
+        let mock_server = wiremock::MockServer::start().await;
+        let template = wiremock::ResponseTemplate::new(http::StatusCode::from($status));
+        let template = template$(.$func($($arg),*))*;
+        wiremock::Mock::given(wiremock::matchers::method("GET")).respond_with(template).mount(&mock_server).await;
+        mock_server
+    }};
 }
 
-pub async fn get_mock_server_with_content<S>(response_code: S, content: Option<&str>) -> MockServer
+pub(crate) async fn get_mock_client_response<T, E>(request: T) -> crate::Response
 where
-    S: Into<StatusCode>,
+    Request: TryFrom<T, Error = E>,
+    ErrorKind: From<E>,
 {
-    let mock_server = MockServer::start().await;
-    let template = ResponseTemplate::new(response_code.into());
-
-    let template = if let Some(s) = content {
-        template.set_body_string(s)
-    } else {
-        template
-    };
-
-    Mock::given(path("/"))
-        .respond_with(template)
-        .mount(&mock_server)
-        .await;
-
-    mock_server
+    ClientBuilder::default()
+        .build()
+        .unwrap()
+        .check(request)
+        .await
+        .unwrap()
 }
 
 /// Helper method to convert a string into a URI
 /// Note: This panics on error, so it should only be used for testing
-pub fn website(url: &str) -> Uri {
-    Uri::Website(Url::parse(url).expect("Expected valid Website URI"))
+pub(crate) fn website(url: &str) -> Uri {
+    Uri::from(Url::parse(url).expect("Expected valid Website URI"))
 }
 
-pub fn mail(address: &str) -> Uri {
-    Uri::Mail(address.to_owned())
+pub(crate) fn mail(address: &str) -> Uri {
+    if address.starts_with("mailto:") {
+        Url::parse(address)
+    } else {
+        Url::parse(&(String::from("mailto:") + address))
```

Comment:
Format could also work here, right?

```
Url::parse(&format!("mailto:{}", address))
```

Reply:
Yes. I just don't like calling `format!`, or anything using format strings like `write!`, unless not doing so considerably reduce readability.

These macros calls `format_args!` and that's not really efficient for simple cases like this.

See the call graph: https://lib.rs/crates/ufmt